custom/icds_reports/management/commands/dashboard_usage_report.py

```diff
@@ -0,0 +1,244 @@
+                                                        time_of_use__lte=end_date)
+                    .annotate(indicator=KeyTextTransform('indicator', 'post_data')).values('indicator',
+                                                                                           'username')
+                    .annotate(count=Count('indicator')).order_by('username', 'indicator'))
+            for record in query:
+                user_counts[record['username']] += record['count']
+                user_indicators[record['username']][int(record['indicator']) - 1] = record['count']
+
+        return user_counts, user_indicators
+
+    def get_role_from_username(self, username):
+        """
+        :param username:
+        :return: Role of the user bases on his username
+        """
+        for key, value in self.roles.items():
+            if key.lower() in username:
+                return value
+        return 'N/A'
+
+    def get_data_for_usage_report(self, date):
+        """
+        :param date: Date for which the report is needed
+        :return:Returns a list of activity report records based on the filters
+        """
+        return list(DashboardUserActivityReport.objects.filter(date=date).values('username', 'state_id',
+                                                                                 'district_id', 'block_id',
+                                                                                 'user_level'))
+
+    def get_location_name_string_from_location_id(self, location_id):
+        return location_id.replace('_id', '_name')
+
+    def get_request_data(self, usernames, user_total_counts, user_indicators_count):
+        first_sheet_headers = ['Sr.No', 'State/UT Name', 'District Name', 'Block Name', 'Username', 'Level',
+                               'Role', 'Total', 'Child', 'Pregnant Women', 'Demographics', 'System Usage',
+                               'AWC infrastructure', 'Child Growth Monitoring List', 'ICDS - CAS Monthly Register',
+                               'AWW Performance Report', 'LS Performance Report', 'Take Home Ration']
+        second_sheet_headers = ['Sr.No', 'State/UT Name',
+                                'No. of times CAS data export downloaded (December 2019)']
+        request_data = list()
+        cas_data = list()
+
+        location_type_filter = {
+            'aggregation_level': 5
+        }
+        for test_location in self.location_test_fields:
+            location_type_filter[test_location] = 0
+        all_awc_locations = AwcLocation.objects.filter(**location_type_filter).values(*self.required_fields)
+
+        location_mapping = self.get_location_id_vs_name_mapping_from_queryset(all_awc_locations)
+        usage_data = list()
+        date = datetime.now()
+        # fetching the last available dashboard activity report
+        while not usage_data:
+            usage_data = self.get_data_for_usage_report(date)
```

Comment:
could we directly use Usage report Class to pull the data like [here](https://github.com/dimagi/commcare-hq/blob/master/custom/icds_reports/tasks.py#L907).
Pulling it directly using this, you will not need to manually do the role mapping here and probably you wont require to deal with locations.

Reply:
running the get_excel_data from DashboardUsage will recalculate everything right.Instead i used Last available DashboardActivityReport to get the username vs location mapping.

Reply:
will change it

Reply:
done